### PR TITLE
Change clock source to CLOCK_MONOTONIC in 'pthread_cond_timedwait' : Update mono-os-mutex.h

### PR DESCRIPTION
**Pull request by jaymin1328**
'Wait' functions for Mutex, Semaphore and Condition (like  "pthread_cond_timedwait") for 'task' synchronisation in Linux uses CLOCK_REALTIME by default, 
this creates many problems where a process changes time. e.g. Command line utility < date -s "desired date and time" >
A thread may be blocked for a more\less time depending upon time change. It is always desirable to use clock source immune to such changes
[http://stackoverflow.com/questions/14248033/clock-monotonic-and-pthread-mutex-timedlock-pthread-cond-timedwait](url)
[http://stackoverflow.com/a/14397906/7042283](url)

[http://stackoverflow.com/questions/3006259/what-time-function-do-i-need-to-use-with-pthread-cond-timedwait](url)
However, it is not possible to change clock source for mutex and semaphore wait functions in Linux. So changing it for condition only

This can be seen by a simple application if you make a simple timer using System.Timers.Timer and print time stamp every regular interval say 2 seconds, now if time shifts in the system for whatever reason ( you can do this in Linux by <sudo date -s "Wed Oct 26 20:25:24 IST 2016"> put desired time in quotes), this print stops coming for the offset time duration. 

The reason behind this is the call to 'pthread_cond_timedwait' function, by Timer class which uses it to sleep for timeout duration. Now this function uses CLOCK_REALTIME which can be changed, so if you change this clock, a thread will be blocked for the unexpected time. So using CLOCK_MONOTONIC removes the error.

There are many embedded systems which do not have RTC and rely upon the external time like systems using GPS or set top box. So this fix is inevitable for them, at least for condition variable wait.
## This is the sample code 

`using System;
using System.Timers;

public class Example
{
    private static System.Timers.Timer aTimer;

```
public static void Main()
{
    SetTimer();

    Console.WriteLine("\nPress the Enter key to exit the application...\n");
    Console.WriteLine("The application started at {0:HH:mm:ss.fff}", DateTime.Now);
    Console.ReadLine();
    aTimer.Stop();
    aTimer.Dispose();

    Console.WriteLine("Terminating the application...");
}

private static void SetTimer()
{
    // Create a timer with a two second interval.
    aTimer = new System.Timers.Timer(2000);
    // Hook up the Elapsed event for the timer. 
    //((System.ComponentModel.ISupportInitialize)(aTimer)).BeginInit();
    aTimer.Elapsed += OnTimedEvent;
    //((System.ComponentModel.ISupportInitialize)(aTimer)).EndInit();
    aTimer.AutoReset = true;
    aTimer.Enabled = true;
}

private static void OnTimedEvent(Object source, ElapsedEventArgs e)
{
    Console.WriteLine("__The Elapsed event was raised at {0:HH:mm:ss.fff}",
                      e.SignalTime);
}
```

}`

### CODE REVIEW

**Comment by jaymin1328** [comment]
@ludovic-henry

**Comment by jaymin1328** [comment]
https://patch-diff.githubusercontent.com/raw/mono/mono/pull/3828.patch

https://patch-diff.githubusercontent.com/raw/mono/mono/pull/3828.diff

**Comment by jaymin1328** [comment]
@akoeplinger   @migueldeicaza @alexrp 
The timers still creates problem and this is different from this issue 
[https://github.com/mono/mono/pull/3414](url)

File: mono/utils/mono-os-mutex.h [hunk 1]
```diff
@@ -118,12 +118,22 @@ mono_os_mutex_unlock (mono_mutex_t *mutex)
 		g_error ("%s: pthread_mutex_unlock failed with \"%s\" (%d)", __func__, g_strerror (res), res);
 }
 
+#if !defined(PLATFORM_MACOSX)	
+	pthread_condattr_t attr;
```
Comment (ludovic-henry) [inline mono/utils/mono-os-mutex.h]: Why is this variable outside the function?

File: mono/utils/mono-os-mutex.h [hunk 2]
```diff
@@ -118,12 +118,22 @@ mono_os_mutex_unlock (mono_mutex_t *mutex)
...
 static inline void
 mono_os_cond_init (mono_cond_t *cond)
 {
 	int res;
-
+/* Attach an attribute having CLOCK_MONOTONIC to condition */	
+#if !defined(PLATFORM_MACOSX)	
+	pthread_condattr_init(&attr);
+	pthread_condattr_setclock(&attr, CLOCK_MONOTONIC);
+	res=pthread_cond_init(cond, &attr);
```
Comment (ludovic-henry) [inline mono/utils/mono-os-mutex.h]: Please keep formatting consistent : space around `=`, space before `(`

File: mono/utils/mono-os-mutex.h [hunk 3]
```diff
@@ -151,22 +161,38 @@ mono_os_cond_wait (mono_cond_t *cond, mono_mutex_t *mutex)
...
+	
+#if defined(PLATFORM_MACOSX)
 	struct timeval tv;
-	struct timespec ts;
 	gint64 usecs;
+#endif	
+	struct timespec ts;
 	int res;
-
+
```
Comment (ludovic-henry) [inline mono/utils/mono-os-mutex.h]: Whitespace noise

File: mono/utils/mono-os-mutex.h [hunk 4]
```diff
@@ -151,22 +161,38 @@ mono_os_cond_wait (mono_cond_t *cond, mono_mutex_t *mutex)
...
 	if (timeout_ms == MONO_INFINITE_WAIT) {
 		mono_os_cond_wait (cond, mutex);
 		return 0;
 	}
 
 	/* ms = 10^-3, us = 10^-6, ns = 10^-9 */
-
+#if !defined(PLATFORM_MACOSX)
+/* cond is using CLOCK_MONOTONIC as time source */	
+	res = clock_gettime (CLOCK_MONOTONIC, &ts);
```
Comment (ludovic-henry) [inline mono/utils/mono-os-mutex.h]: You shouldn't use the `g_error` from below, as the message is specifically for an error in `gettimeofday`. So please duplicate the error checking code.

File: mono/utils/mono-os-mutex.h [hunk 5]
```diff
@@ -151,22 +161,38 @@ mono_os_cond_wait (mono_cond_t *cond, mono_mutex_t *mutex)
...
+	
 	if (timeout_ms == MONO_INFINITE_WAIT) {
 		mono_os_cond_wait (cond, mutex);
 		return 0;
 	}
 
 	/* ms = 10^-3, us = 10^-6, ns = 10^-9 */
-
+#if !defined(PLATFORM_MACOSX)
+/* cond is using CLOCK_MONOTONIC as time source */
```
Comment (ludovic-henry) [inline mono/utils/mono-os-mutex.h]: Indentation

File: mono/utils/mono-os-mutex.h [hunk 6]
```diff
@@ -151,22 +161,38 @@ mono_os_cond_wait (mono_cond_t *cond, mono_mutex_t *mutex)
...
 		return 0;
 	}
 
 	/* ms = 10^-3, us = 10^-6, ns = 10^-9 */
-
+#if !defined(PLATFORM_MACOSX)
+/* cond is using CLOCK_MONOTONIC as time source */	
+	res = clock_gettime (CLOCK_MONOTONIC, &ts);
+#else
+/* clock_gettime is not supported in MAC OS x */
```
Comment (ludovic-henry) [inline mono/utils/mono-os-mutex.h]: Identation

File: mono/utils/mono-os-mutex.h [hunk 7]
```diff
@@ -118,12 +118,22 @@ mono_os_mutex_unlock (mono_mutex_t *mutex)
...
+#if !defined(PLATFORM_MACOSX)	
+	pthread_condattr_t attr;
+#endif
+
 static inline void
 mono_os_cond_init (mono_cond_t *cond)
 {
 	int res;
-
+/* Attach an attribute having CLOCK_MONOTONIC to condition */
```
Comment (ludovic-henry) [inline mono/utils/mono-os-mutex.h]: Identation

File: mono/utils/mono-os-mutex.h (hunk 1 above)
Comment (jaymin1328) [inline mono/utils/mono-os-mutex.h]: w32handle.c has used static mono_cond_t so , I had applied the same logic here. but now it is local

Comment (jaymin1328) [inline mono/utils/mono-os-mutex.h]: w32handle.c has used static mono_cond_t so , I had applied the same logic here. but now it is local

**Comment by ludovic-henry** [comment]
build

**Comment by BrandonLWhite** [comment]
I really need this bug fixed in 4.8.

**Comment by jaymin1328** [comment]
@BrandonLWhite : u can safely use my code, its working fine

**Comment by luhenry** [comment]
This has been merged with https://github.com/mono/mono/pull/4446

**Pull request closed without merging**